### DISCUSSION TITLE
BRIL for GPUs

### DISCUSSION THREAD

**Discussion by chsasank** [sampsyo/bril#316] [Ideas]
Hi

I am working on some sort of IR that unifies execution of GPUs from all three major vendors. Please [checkout my critique](https://twitter.com/sasank51/status/1776188399033975126) on LLVM on why I am not happy with it. My ideal solution is already somewhat implemented by [Aksel Alpay](https://emcl.iwr.uni-heidelberg.de/people/alpay-aksel) from Hidelberg university. Please see [his paper](https://dl.acm.org/doi/abs/10.1145/3585341.3585351) on single pass sycl compiler. I am looking to formalize the interface to his runtime JIT compiler into an IR.

I have considered my options for this IR and I feel BRIL might be the right fit for me for the following reasons:

1. JSON or XML is amazing because I can interface it with any language. Any language can be used for implementation instead of just C/C++.
2. SSA is painful to represent in such a text/human readable format. If I have to give up on one of readability or SSA, I will give up on SSA and accept three address code (TAC).
3. Hackability and getting started is more important than 'production' grade. Your educational material helps people get started easily.

BRIL like design fits these constraints! It's too much work to completely give up on LLVM yet. I'll just use BRIL as sort of a frontend to LLVM like one of the projects your students have on the BRIL repo.

What do you think?

**Comment by chsasank** [comment]
I have documented more of my thoughts on implementation here: https://outline.von-neumann.ai/s/71b5caeb-c307-4e58-bf83-67cfda3113ef

**Reply by sampsyo** [reply]
This seems super fun! Thanks for the extra details here. I like the idea of an S-expression form that can generate Bril.

The big thing that I'm curious about at this point is: what language extensions are necessary to make Bril into a practical GPU language? This probably sounds a little basic, but I guess I'm curious to start at that basic level… i.e., to think about what makes a GPU IR a GPU IR. Are there specific new instructions that we'd need? It seems like memory instructions might be paramount? (Or anything beyond instructions that I'm not thinking of?)

**Reply by chsasank** [reply]
I have written down my thoughts on GPU IR here: https://outline.von-neumann.ai/s/d0cd5eb9-2e15-4fa4-bd17-c3911f305008/doc/the-plan-for-gpu-ir-IgRr11a1yC

Please have a look.

**Reply by sampsyo** [reply]
That's helpful; thanks! It sounds like you want to start by including _both_ host-side and GPU-side code, which is interesting. All this stuff makes sense.

The next step I'd be interested in is thinking through the concrete extensions to the Bril language that would be required. Specifically, we should probably start with a list of new Bril instructions? Your post has a list of LLVM intrinsics (e.g., `tid.x` when targeting PTX)… do you imagine replicating these in Bril? Regardless, sketching out the list of opcodes (maybe in the same general form as [the docs for other Bril language extensions](https://capra.cs.cornell.edu/bril/lang/memory.html)) would be a great next step.
